rfcs/v3/authorization-rules.md

```diff
@@ -0,0 +1,568 @@
+- Composability: The ability to define multiple simple rules that compose automatically instead of one big expression.
+- Performance: The time it takes to evaluate the permissions for a particular GraphQL request should only depend on the metadata objects being queried and not on the total size of the metadata.
+- Validation: We should be able to statically validate any references to the entities in the metadata (eg: fields, relationships, commands, etc.).
+
+For a given resource (type / model / command), instead of defining permissions for a given role, you would define multiple rules, where each rule contains:
+- One or more composable permission primitives corresponding to the resource.
+- A condition which must evaluate to true at runtime for this rule to apply.
+
+The final permissions for that resource are a composition of the permission primitives defined across rules whose condition evaluated to true.
+
+### Condition
+
+The condition that may be attached to a rule is a boolean expression built using the following primitives.
+
+#### Boolean Operators
+
+The condition must start at a boolean operator. These are some potential boolean operators:
+- `and`: array of boolean expressions which are ANDed together.
+- `or`: array of boolean expressions which are ORed together.
+- `not`: boolean exrepssion which is negated.
+- `equal`: Compares two value expressions are the same.
+- `isNull`: whether a given value is null
+- `contains`: Whether a given value is contained within a given array of values
+- `greaterThan`: Compares left > right.
+- `lessThan`: Compares left < right.
+- `greaterThanOrEqual`: Compares left >= right
+- `lessThanOrEqual`: Compares left <= right
+- `regexMatch`: Whether a given value matches the given regular expression
+
+#### Value Expressions
+
+The left and right side of a comparison expression are value expressions which can either be:
+- Literals
+- Session Variable references
+
+In the future, these may be expanded to:
+- The output from an invoked command
+- The result of invoking an external policy engine like Open Policy Agent (OPA).
+
+#### Example
+
+Here is a rule that descriebs the condition to enable an authorization rule if the user is an admin or is an unbanned user in the PRO tier.
+
+```yaml
+or:
+  - and:
+    - equal:
+        left:
+          sessionVariable: customer_tier
+        right:
+          literal: PRO_TIER
+    - not:
+        equal:
+          left:
+            sessionVariable: is_banned
+          right:
+            literal: true
+  - equal:
+      left:
+        sessionVariable: is_admin
+      right:
+        literal: true
+```
+
+### Type Permissions
+
+The following permission primitives are available when defining a permissions rule for a type.
+- Allow access to certain fields
+- Deny access to certain fields
+
+If a field is both allowed and denied, it will be denied.
+
+#### Example
+
+```yaml
+kind: TypePermissions
+version: v2
+definition:
```

Comment:
What do rules like these mean for the generated GraphQL schema?

Will `internal_order_notes` always be visible, but we'll make it nullable?

Reply:
(or would accessing a field that one is not allowed trigger an error instead of a `null` field?)

Reply:
`internal_order_notes` will not be visible in the GraphQL schema if the `x-hasura-role` session variable is not set to `admin` in the introspection query.

(Note that there is a separate conversation going on around eliminating dynamic role based schemas altogether but for the purpose of this RFC, let's assume we still want the GraphQL schema to contain only what's accessible)